Move away from ReplaySubject's when sending `self`

sprynmr opened the pull request:
Since `self` was being sent to the subjects with a buffer of 1, it was
holding itself in the subjects indefinitely. The weakSelf I added in #30
only affects the closure itself, not the memory semantics of the subject
storing itself.

Looking back at RVMViewModel, they didn't have these as a replay
subject, so I changed them to publish subjects to match their semantics.

The throttling observable can keep it's subject since it's not sending
values of self.

I *think* this is correct, but please correct me on any of this. :)

sprynmr commented:
Consumers of the observables should do replay themselves if desired, as they can tie this lifecycle to their own.

Alternatively, maybe we could use the share semantic, where the buffer clears when the subscriber count goes to zero?

sprynmr commented:
More accurately, `shareReplayLatestWhileConnected` behavior.

esttorhe commented:
Awesome job Bob!! 😄 

Going with the `share` behavior sounds like the correct way to go actually.


QQ. Would you mind running the tests and if they pass merging this?

I seriously need to fix the `CI` 😖

sprynmr commented:
Yep, looking into share and running tests now. Just have to get the right version of cocoapods to install the dependencies.

esttorhe commented:
```
make bootstrap

make bundle
```

😉

sprynmr commented:
I switched away from the manual subjects and just chain off of the
`activeObservable` with filtering. I *believe* `unowned` is ok here so
that we can leave the observable non-optional, since the only way the
`activeObservable` will fire a value is by something being set on
`self`, making that closure mutually dependent on self existing.

Removed the import of RxCocoa in the tests since it's not used anymore.
Tests still sporadically fail, I think it has something to do with the
interval timer, but not 100% sure.

sprynmr commented:
Hmm. So there's still an issue with the `shareReplayLatestWhileConnected` when the view model is subscribing to it's own `didBecomeActive` observable, unless you add some `take` mechanic (like take 1). Otherwise the subscriber count will never go to zero, and then the subject will keep `self` alive.

```
        let _ = self.didBecomeActive
            .take(1) // without this, retain cycle
            .subscribeNext { [weak self] _ in
                self?.context.fetchUpgradeSlotsIfNeeded()
            }
```

This is common enough in my code that I think we should just move to publish subjects. There's not really much gained here by having replay I don't think. If it's needed, a consumer can replay *outside* of the view model and tie it to a lifecycle of say a VC.

sprynmr commented:
Recommend we ditch trying to share the observable chain since nothing
terribly expensive is going on there, and having replay semantics causes
retain cycles if the consumer subscribes to the `didBecomeActive`
observable in the view model itself (and doesn't maintain the lifecycle
with `take(1)` or similar.) Either that, or we can move away from the
value sent being the view model. I don't have any instances where I
subscribe to the `didBecomeActive` observable after it's already active.

esttorhe commented:
I think we should move away from the value sent being the `VM` (given that it doesn't actually add much value) and also would potentially simplify the fix.

Also loved how you simplify the code.

Been meaning to refactor this for quite some time now and this is way better than what i thought of at first 🎉 

If you feel confident enough go ahead and merge and i'll release a new pod :)

sprynmr commented:
Is there a way to merge this without the circleci stuff passing? I guess we should disable that for now.

esttorhe commented:
Yes; give me a sec.

Defo We should disable it for now.

I contacted them once; I need to send a followup ✉️ so they can re-enable de `iOS` instance.

Still undecided if I should do that or move this to `Travis` and be done with this.

esttorhe commented:
@sprynmr merged, `3.1.2` tagged and released and a new `pod` was pushed to trunk!!

Thanks again!!!
